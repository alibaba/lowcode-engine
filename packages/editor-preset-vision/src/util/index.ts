function getUrlParameter(sParam: string) {
  var sPageURL = window.location.search.substring(1),
    sURLVariables = sPageURL.split('&'),
    sParameterName,
    i;

  for (i = 0; i < sURLVariables.length; i++) {
    sParameterName = sURLVariables[i].split('=');

    if (sParameterName[0] === sParam) {
      return sParameterName[1] === undefined ? true : decodeURIComponent(sParameterName[1]);
    }
  }
};

export function getFormUuid() {
  return getUrlParameter('formUuid');
}