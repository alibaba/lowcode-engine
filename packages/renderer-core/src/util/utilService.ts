import {
  type AnyFunction,
  type UtilDescription,
  createDecorator,
  type StringDictionary,
} from '@alilc/lowcode-shared';
import { isPlainObject } from 'lodash-es';
import { IPackageManagementService } from '../package';
import { ICodeRuntimeService } from '../code-runtime';

export interface IRuntimeUtilService {
  add(utilItem: UtilDescription, force?: boolean): void;
  add(name: string, target: AnyFunction | StringDictionary, force?: boolean): void;

  remove(name: string): void;
}

export const IRuntimeUtilService = createDecorator<IRuntimeUtilService>('rendererUtilService');

export class RuntimeUtilService implements IRuntimeUtilService {
  private _utilsMap: Map<string, any> = new Map();

  constructor(
    utils: UtilDescription[] = [],
    @ICodeRuntimeService private codeRuntimeService: ICodeRuntimeService,
    @IPackageManagementService private packageManagementService: IPackageManagementService,
  ) {
    for (const util of utils) {
      this.add(util);
    }
    this._injectScope();
  }

  add(utilItem: UtilDescription, force?: boolean): void;
  add(name: string, fn: AnyFunction | StringDictionary, force?: boolean): void;
  add(
    util: UtilDescription | string,
    fn?: AnyFunction | StringDictionary | boolean,
    force?: boolean,
  ): void {
    let name: string;
    let utilObj: AnyFunction | StringDictionary | UtilDescription;

    if (typeof util === 'string') {
      if (!fn) return;

      name = util;
      utilObj = fn as AnyFunction | StringDictionary;
    } else {
      if (!util) return;

      name = util.name;
      utilObj = util;
      force = fn as boolean;
    }

    this._addUtilByName(name, utilObj, force);
  }

  private _addUtilByName(
    name: string,
    fn: AnyFunction | StringDictionary | UtilDescription,
    force?: boolean,
  ): void {
    if (this._utilsMap.has(name) && !force) return;

    if (isPlainObject(fn)) {
      if ((fn as UtilDescription).type === 'function' || (fn as UtilDescription).type === 'npm') {
        const utilFn = this._parseUtil(fn as UtilDescription);
        if (utilFn) {
          this._addUtilByName(name, utilFn, force);
        }
      } else if ((fn as StringDictionary).destructuring) {
        for (const key of Object.keys(fn)) {
          this._addUtilByName(key, (fn as StringDictionary)[key], force);
        }
      } else {
        this._utilsMap.set(name, fn);
      }
    } else if (typeof fn === 'function') {
      this._utilsMap.set(name, fn);
    }
  }

  remove(name: string): void {
    this._utilsMap.delete(name);
  }

  private _parseUtil(utilItem: UtilDescription) {
    if (utilItem.type === 'function') {
      return this.codeRuntimeService.rootRuntime.run(utilItem.content.value);
    } else {
      return this.packageManagementService.getModuleByReference(utilItem.content);
    }
  }

  private _injectScope(): void {
    const exposed = new Proxy(Object.create(null), {
      get: (_, p: string) => {
        return this._utilsMap.get(p);
      },
      set() {
        return false;
      },
      has: (_, p: string) => {
        return this._utilsMap.has(p);
      },
    });

    this.codeRuntimeService.rootRuntime.getScope().set('utils', exposed);
  }
}
