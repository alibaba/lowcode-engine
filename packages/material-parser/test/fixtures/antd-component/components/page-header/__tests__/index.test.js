import React from 'react';
import { mount, render } from 'enzyme';
import PageHeader from '..';
import ConfigProvider from '../../config-provider';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { spyElementPrototypes } from '../../__tests__/util/domHook';

describe('PageHeader', () => {
  mountTest(PageHeader);
  rtlTest(PageHeader);

  let spy;

  beforeAll(() => {
    spy = spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        width: 100,
      }),
    });
  });

  afterAll(() => {
    spy.mockRestore();
  });

  it('pageHeader should not contain back it back', () => {
    const routes = [
      {
        path: 'index',
        breadcrumbName: 'First-level Menu',
      },
      {
        path: 'first',
        breadcrumbName: 'Second-level Menu',
      },
      {
        path: 'second',
        breadcrumbName: 'Third-level Menu',
      },
    ];
    const wrapper = mount(<PageHeader title="Page Title" breadcrumb={{ routes }} />);
    expect(wrapper.find('.ant-page-header-back')).toHaveLength(0);
  });

  it('pageHeader should have breadcrumb', () => {
    const routes = [
      {
        path: 'index',
        breadcrumbName: 'First-level Menu',
      },
    ];
    const wrapper = mount(<PageHeader title="Page Title" breadcrumb={{ routes }} />);
    expect(wrapper.find('.ant-breadcrumb')).toHaveLength(1);
    expect(wrapper.find('.ant-page-header-back')).toHaveLength(0);
  });

  it('pageHeader should no contain back', () => {
    const wrapper = mount(<PageHeader title="Page Title" backIcon={false} />);
    expect(wrapper.find('.ant-page-header-back')).toHaveLength(0);
  });

  it('pageHeader should contain back it back', () => {
    const callback = jest.fn(() => true);
    const wrapper = mount(<PageHeader title="Page Title" onBack={callback} />);
    expect(wrapper.find('.ant-page-header-back')).toHaveLength(1);
  });

  it('pageHeader onBack transfer', () => {
    const callback = jest.fn(() => true);
    const wrapper = mount(<PageHeader title="Page Title" onBack={callback} />);
    wrapper.find('div.ant-page-header-back-button').simulate('click');
    expect(callback).toHaveBeenCalled();
  });

  it('pageHeader should support className', () => {
    const wrapper = render(
      <PageHeader title="Page Title" className="not-works" backIcon={false} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('pageHeader should not render blank dom', () => {
    const wrapper = render(<PageHeader title={false} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('breadcrumbs and back icon can coexist', () => {
    const routes = [
      {
        path: 'index',
        breadcrumbName: 'First-level Menu',
      },
      {
        path: 'first',
        breadcrumbName: 'Second-level Menu',
      },
      {
        path: 'second',
        breadcrumbName: 'Third-level Menu',
      },
    ];
    const wrapper = mount(<PageHeader title="Title" breadcrumb={{ routes }} />);
    expect(wrapper.find('.ant-breadcrumb')).toHaveLength(1);

    wrapper.setProps({ onBack: () => {} });
    expect(wrapper.find('.ant-breadcrumb')).toHaveLength(1);
  });

  it('pageHeader should render correctly int RTL direction', () => {
    const wrapper = mount(
      <ConfigProvider direction="rtl">
        <PageHeader title="Page Title" />
      </ConfigProvider>,
    );

    expect(render(wrapper)).toMatchSnapshot();
  });

  it('change container width', () => {
    const wrapper = mount(<PageHeader title="Page Title" extra="extra" />);
    wrapper.triggerResize();
    wrapper.update();
    expect(wrapper.find('.ant-page-header').hasClass('ant-page-header-compact')).toBe(true);
  });
});
