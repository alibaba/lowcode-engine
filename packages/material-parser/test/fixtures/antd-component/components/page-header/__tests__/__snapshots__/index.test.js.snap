// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader pageHeader should not render blank dom 1`] = `
<div
  class="ant-page-header ant-page-header-ghost"
/>
`;

exports[`PageHeader pageHeader should render correctly int RTL direction 1`] = `
<div
  class="ant-page-header ant-page-header-ghost ant-page-header-rtl"
>
  <div
    class="ant-page-header-heading"
  >
    <div
      class="ant-page-header-heading-left"
    >
      <span
        class="ant-page-header-heading-title"
        title="Page Title"
      >
        Page Title
      </span>
    </div>
  </div>
</div>
`;

exports[`PageHeader pageHeader should support className 1`] = `
<div
  class="ant-page-header not-works ant-page-header-ghost"
>
  <div
    class="ant-page-header-heading"
  >
    <div
      class="ant-page-header-heading-left"
    >
      <span
        class="ant-page-header-heading-title"
        title="Page Title"
      >
        Page Title
      </span>
    </div>
  </div>
</div>
`;

exports[`PageHeader rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-page-header ant-page-header-ghost ant-page-header-rtl"
/>
`;
